### PR TITLE
Automatically load and store cache hierarchy when constructing FusionCache

### PR DESCRIPTION
**Pull request by rdspring1**
This PR enables loading and storing the fusion cache hierarchy automatically from a common workspace.

When constructing the `FusionCache`, we deserialize from `tmp/nvfuser_kernel_db/nvf_serde`. This step can be disabled by passing a bool flag to `FusionCache::get`.

In python, we register with `atexit` the `FusionCache::serialize` function to automatically serialize when exiting the program.

```python
# Use atexit to automatically call serialize on program exit
import atexit
import nvfuser
atexit.register(nvfuser.serialize)
```

Serialization saves the `FusionCache` to a per-process temporary file first to avoid conflicts with multiple processes. Then, the temporary file is renamed to the actual file. If the actual file already exists, then the `std::filesystem::rename` call may fail or replace the actual file. This approach is taken from PyTorch Jiterator.

The saved workspace binary may become out-of-date with the NvFuser repository. If the binary cannot be verified, it is deleted and the `FusionCache` is not restored.
 
Reference: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/jit_utils.cpp#L1602-L1629

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: python_tests/test_python_frontend.py
Comment (rdspring1) [reply python_tests/test_python_frontend.py:23]: I wanted the call to be `nvfuser.serialize`, which would distinguish it from `fusion_cache.serialize`. I thought that `serialize` is a common function name and there could be potential name conflicts.

File: csrc/python_frontend/fusion_cache.cpp
Comment (rdspring1) [reply csrc/python_frontend/fusion_cache.cpp:21]: Unlikely. The code comes from PyTorch Jiterator, so it should be reliable.

File: python_tests/test_python_frontend.py
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:23]: Got ya. I'm fine with it as-is, since this is only testing scripts.

Alternatively, something with this can avoid conflicts. i.e.
```
from nvfuser import (
  serialize as nv_serialize,
)
```

Comment (rdspring1) [reply python_tests/test_python_frontend.py:23]: Good idea! I forgot you could do that in python.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: csrc/python_frontend/fusion_cache.cpp
Comment (jjsjann123) [inline csrc/python_frontend/fusion_cache.cpp:337]: nitpick: maybe throw a warning here when we are removing things.
I'm also wondering, should this be put into a try block? e.g. if we try to remove something that we can't write to.

**Review by jjsjann123 — COMMENTED** [review]
What's with the segfault in the CI?

**Review by jjsjann123 — APPROVED** [review]
code looks mostly fine with me. I'm stamping just to unblock the merge.
Having said that, please do make sure that we have a green CI before merging. :bow:

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
@kevinstephano Looks like there was an issue with #1313. The `FullOp` now uses `TensorCreationSymbolic` in `fusion_record_serde.cpp`

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Pull request merged**
